src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,181 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="BulkUpdateGraph.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.Client
+{
+    /// <summary>
+    /// Contains information about descriptors and their related descriptors. 
+    /// </summary>
+    internal class BulkUpdateGraph
+    {
+        /// <summary>Set of related descriptors</summary>
+        private Dictionary<Descriptor, HashSet<Descriptor>> relatedDescriptors;
+
+        /// <summary>Set of top level descriptors</summary>
+        private List<Descriptor> topLevelDescriptors;
+
+        /// <summary> The EntitySetName of the top-level descriptor types </summary>
+        private string entitySetName;
+
+        /// <summary>where to pull the changes from</summary>
+        protected readonly RequestInfo requestInfo;
+
+        internal BulkUpdateGraph(RequestInfo requestInfo)
+        { 
+            this.relatedDescriptors= new Dictionary<Descriptor, HashSet<Descriptor>>(); 
+            this.topLevelDescriptors= new List<Descriptor>();
+            this.requestInfo = requestInfo;
+        }
+
+        /// <summary>
+        /// Returns a collection of all the root descriptors used in a deep update.
+        /// </summary>
+        internal List<Descriptor> TopLevelDescriptors
+        {
+            get
+            {
+                return this.topLevelDescriptors;
+            }
+        }
+
+        /// <summary>
+        /// Returns the EntitySetName.
+        /// </summary>
+        internal string EntitySetName
+        {
+            get
+            {
+                return this.entitySetName;
+            }
+        }
+
+        /// <summary>
+        /// Computes the EntitySetName from the given entity.
+        /// </summary>
+        internal void SetEntitySetName()

```

Comment:
Set... , basically, takes an input to change the state of the instance. What's it for?

Reply:
It sets the `EntitySetName`. In some cases the `EntitySetName` name of the top-level `EntityDescriptor` might not be set and needs to be found to be used when calling the `CreateODataDeltaResourceSetWriter`. In that case, we need to find the `EntitySetName` manually from the `Entity`.

Reply:
or I should change the name to `ComputeEntitySetName`